File: 383. Ransom Note/383. Ransom Note.java
import java.util.HashMap;

class Solution {

  public boolean canConstruct(String ransomNote, String magazine) {
    HashMap<Character, Integer> letters = new HashMap<>();

    for (char c : magazine.toCharArray()) {
      letters.put(c, (letters.getOrDefault(c, 0)) + 1);
    }

    for (char c : ransomNote.toCharArray()) {
      if (letters.getOrDefault(c, 0) <= 0) {
        return false;
      }

      letters.put(c, letters.get(c) - 1);
    }

    return true;
  }
}

File: 383. Ransom Note/383. Ransom Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  const letters = {};

  for (const c of magazine) {
    letters[c] = (letters[c] || 0) + 1;
  }

  for (const c of ransomNote) {
    if (!letters[c] || letters[c] <= 0) {
      return false;
    }

    letters[c]--;
  }

  return true;
};
